xpdan/data_reduction.py

```diff
@@ -0,0 +1,360 @@
+            else:
+                pass
+        # get data fields
+        for key in self.data_fields:
+            val = event['data'].get(key, None)
+            if el is not None:
+                feature = "{}={}".format(key, val)
+                feature_list.append(feature)
+            else:
+                pass
+        # get uid
+        feature_list.append(uid)
+        return "_".join(feature_list)
+
+    def pull_dark(self, header):
+        dark_uid = header.start.get(an_glbl.dark_field_key, None)
+        if dark_uid is None:
+            print("INFO: no dark frame is associated in this header, "
+                  "subrraction will not be processed")
+            return None
+        else:
+            dark_search = {'group': 'XPD', 'uid': dark_uid}
+            dark_header = db(**dark_search)
+            dark_img = np.asarray(get_images(dark_header,
+                                             self.image_field)).squeeze()
+        return dark_img, dark_header[0].start.time
+
+    def _dark_sub(self, event, dark_img):
+        """ priviate method operates on event level """
+        dark_sub = False
+        img = event['data'][self.image_field]
+        if dark_img is not None and isinstance(dark_img, np.ndarray):
+            dark_sub = True
+            img -= dark_img
+        ind = event['seq_num']
+        event_timestamp = event['timestamps'][self.image_field]
+        return img, event_timestamp, ind, dark_sub
+
+    def dark_sub(self, header):
+        """ public method operates on header level """
+        img_list = []
+        timestamp_list = []
+        dark_img, dark_time_stamp = self.pull_dark(header)
+        for ev in get_events(header, fill=True):
+            sub_img, timestamp, ind, dark_sub = self._dark_sub(ev, dark_img)
+            img_list.append(sub_img)
+            timestamp_list.append(timestamp)
+        return img_list, timestamp_list, dark_img, header.start
+
+    def _file_name(self, event, event_timestamp, ind):
+        """ priviate method operates on event level """
+        f_name = self._feature_gen(event)
+        f_name = '_'.join([f_name,
+                           _timestampstr(event_timestamp)])
+        f_name = '{}_{:04d}.tif'.format(f_name, ind)
+        return f_name
+
+
+# init
+xpd_data_proc = DataReduction()
+ai = AzimuthalIntegrator()
+
+
+### analysis function operates at header level ###
+def _prepare_header_list(headers):
+    if type(list(headers)[1]) == str:
+        header_list = list()
+        header_list.append(headers)
+    else:
+        header_list = headers
+    return header_list
+
+def _load_config():
+    with open(os.path.join(an_glbl.config_base, an_glbl.calib_config_name)) as f:
+        config_dict = yaml.load(f)
+    return config_dict
+
+def _npt_cal(config_dict):
+    """ config_dict should be a PyFAI calibration dict """
+    x_0, y_0 = (config_dict['centerX'], config_dict['centerY'])
+    dist = np.sqrt((2048-x_0)**2 + (2048-y_0)**2)
```

Comment:
I am not certain what the exact answer should be. I examined the standard deviation of each ring, and found that the pixel radial resolution (for flat detectors) gave the standard deviation with the least oscillations in the standard deviation.

I think this is a symptom of a larger problem, namely: what is the best number and size of the bins for integrating the image? I don't have a solid answer yet. As you say, the radial pixel resolution binning makes physical sense, but can be subpar (although I would be interested in examining the results you have for the number of points study).

Personally, I think that all this should be done in Q space, since that will help also address issues of detector rotation and such, but I don't have a mathematical argument for how the bins should be sized in Q space. My previous attempts did not take into account the tilt and rotation of the detector which could be very important to getting the best results.